### PR TITLE
[onert] Plan non-const tensors on train backend

### PR DESCRIPTION
**Pull request by ragmani**
This commit add planning non-const tensors(forwarding non-const tensors) on train backend.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

File: runtime/onert/backend/train/TensorPlanner.cc [hunk 1]
```diff
@@ -35,9 +35,184 @@ TensorPlanner::TensorPlanner(const ir::train::TrainableGraph &tgraph,
...
+      assert(uses_map.find(input_index) != uses_map.end());
+      assert(uses_map[input_index] > 0);
+      uses_map[input_index]--;
+      if (uses_map[input_index] == 0)
+      {
+        // plan for deallocation of static tensor node
+        tensor_builder->notifyLastUse(input_index.index());
+      }
+    }
+  }
```
Comment (ragmani) [inline runtime/onert/backend/train/TensorPlanner.cc:159]: This code is similar to https://github.com/Samsung/ONE/blob/19c259d2b462bf260addc4a24f0b3590cf33c1cf/runtime/onert/core/include/backend/basic/BackendContextHelpers.h#L103-L163

File: runtime/onert/backend/train/TensorPlanner.cc [hunk 2]
```diff
@@ -35,9 +35,184 @@ TensorPlanner::TensorPlanner(const ir::train::TrainableGraph &tgraph,
...
+    for (const auto &output : op_outputs)
+    {
+      if (_external_operands.contains(output))
+        continue;
+      if (!tensor_builder->isRegistered(output))
+        continue;
+
+      const auto output_index = ir::train::TrainingOperandIndex{output, true};
+      assert(defs_map.find(output_index) != defs_map.end());
+      defs_map[output_index] = 0;
```
Comment (jyoungyun) [inline runtime/onert/backend/train/TensorPlanner.cc:113]: (Q) What is the meaning of zero?

File: runtime/onert/backend/train/TensorPlanner.cc [hunk 3]
```diff
@@ -35,9 +35,184 @@ TensorPlanner::TensorPlanner(const ir::train::TrainableGraph &tgraph,
...
 
-void TensorPlanner::planNonConstTensors(TensorBuilder *)
+void TensorPlanner::planNonConstTensors(TensorBuilder *tensor_builder)
 {
-  // TODO Plan non-const tensors
+  VERBOSE(BackendContext) << "Start planning non-constant tensors" << std::endl;
+
+  const auto &training_usedefs = _tgraph.trainingUseDefs();
+
+  // NOTE The uses_map and defs_map must have size of only registered tensors
```
Comment (zetwhite) [inline runtime/onert/backend/train/TensorPlanner.cc:44]: (optional) 

```suggestion
  // NOTE The uses_map and defs_map must have the size of only registered tensors
```

**Review by zetwhite — DISMISSED** [review]
LGTM

File: runtime/onert/backend/train/TensorPlanner.cc (hunk 2 above)
Comment (ragmani) [reply runtime/onert/backend/train/TensorPlanner.cc:113]: The zero means forwarding operand has only one def. The above assertion checks it.

Comment (ragmani) [reply runtime/onert/backend/train/TensorPlanner.cc:113]: The assertion may not be enough to explain that meaning. I'll try add the assertion, `assert(defs_map[output_index] == 1)`.

**Review by jyoungyun — APPROVED** [review]
LGTM

**Review by zetwhite — APPROVED** [review]

**Pull request merged**